### PR TITLE
Factor up dotnet restore in new action, add Qodana action for dotnet

### PR DESCRIPTION
**Pull request by monsieurleberre**
- upgrade to checkout v4
- new Qodana common action aiming to remove issue due to private packages
- still could refactor deployment actions but this already grew too much

### CODE REVIEW

File: restore-dotnet/action.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                  --username "trakx-bot"
+                  --password ${{inputs.packageReadonlyPat}}
+                  --store-password-in-clear-text
+
+    - name: Restore Cache
+      uses: actions/cache@v3
+      with:
+        path: ~/.nuget/packages
+        key: ${{ runner.os }}-nuget-${{ hashFiles('**/packages.lock.json') }}
+        restore-keys: ${{ runner.os }}-nuget-
```
Comment (joaofbvieira) [inline restore-dotnet/action.yml:45]: This is a pre-existent issue, but I think this cache key is missing something 🤣

File: test-dotnet/action.yml [hunk 2]
```diff
@@ -21,47 +21,29 @@ inputs:
...
-      env:
-        DOTNET_INSTALL_DIR: "./.dotnet"  # Required by the runner to install dotnet
-      uses: actions/setup-dotnet@v3
+  
+    - name: Checkout trakx/github-actions repo
+      uses: actions/checkout@v4
       with:
-        dotnet-version: ${{ inputs.dotnetVersion }}
+        repository: trakx/github-actions
+        path: ./github-actions-build-push-container-image
```
Comment (joaofbvieira) [inline test-dotnet/action.yml:38]: I'd suggest to pick a path that is "more less" the current action name.
In this case would be something like:
```
path: ./github-actions-test-dotnet
```

This avoid to have conflicts when we are reusing actions. At least I've struggled a lot due to this when creating shared workflows 😞

**Review by joaofbvieira — COMMENTED** [review]
Nice! 🤩

File: test-dotnet/action.yml (hunk 2 above)
Comment (monsieurleberre) [reply test-dotnet/action.yml:38]: It is a mistake, thanks for spotting it

**Comment by monsieurleberre** [comment]
Seems to work as per https://github.com/trakx/messari-api-client/pull/172

File: restore-dotnet/action.yml (hunk 1 above)
Comment (goneneren) [reply restore-dotnet/action.yml:45]: @monsieurleberre did you checked if restore is working when you change packages?

**Review by goneneren — APPROVED** [review]

**Pull request merged**